plutus-chain-index/app/Indexer.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE ExplicitForAll #-}
+{-# LANGUAGE GADTs          #-}
+{-# LANGUAGE LambdaCase     #-}
+{-# LANGUAGE TupleSections  #-}
+
+module Main where
+
+import Cardano.Api (Block (Block), BlockHeader (BlockHeader), BlockInMode (BlockInMode), BlockNo (BlockNo), CardanoMode,
+                    ChainPoint (ChainPoint), Hash, NetworkId (Mainnet), SlotNo (SlotNo), chainPointToSlotNo,
+                    deserialiseFromRawBytesHex, proxyToAsType)
+import Cardano.Api qualified as C
+import Cardano.BM.Trace (nullTracer)
+import Cardano.Index.Datum (DatumIndex)
+import Cardano.Index.Datum qualified as Ix
+import Cardano.Protocol.Socket.Client (ChainSyncEvent (Resume, RollBackward, RollForward), runChainSync)
+import Control.Concurrent (threadDelay)
+import Control.Lens.Operators ((^.))
+import Control.Monad (forever, when)
+import Data.IORef (IORef, newIORef, readIORef, writeIORef)
+import Data.List (findIndex)
+import Data.Map (assocs)
+import Data.Maybe (fromJust, fromMaybe)
+import Data.Proxy (Proxy (Proxy))
+import Data.Text (pack)
+import Data.Text.Encoding (encodeUtf8)
+import Index.VSplit qualified as Ix
+import Ledger.TimeSlot (SlotConfig (..))
+import Options.Applicative (Parser, execParser, fullDesc, header, help, helper, info, long, metavar, progDesc,
+                            strOption, (<**>))
+import Plutus.ChainIndex.Tx (ChainIndexTx (..))
+import Plutus.Contract.CardanoAPI (fromCardanoTx)
+import Plutus.Script.Utils.V1.Scripts (Datum, DatumHash)
+
+{- | This executable is meant to exercise a set of indexers (for now datumhash -> datum)
+     against the mainnet (meant to be used for testing).
+
+     In case you want to access the results of the datumhash indexer you need to query
+     the resulting database:
+     $ sqlite3 datums.sqlite
+     > select slotNo, datumHash, datum from kv_datumhsh_datum where slotNo = 39920450;
+     39920450|679a55b523ff8d61942b2583b76e5d49498468164802ef1ebe513c685d6fb5c2|X(002f9787436835852ea78d3c45fc3d436b324184
+-}
+
+-- Options
+data Options = Options
+  { socketPath :: FilePath
+  , dbPath     :: FilePath
+  }
+
+options :: Parser Options
+options = Options
+      <$> strOption
+          ( long "socket"
+         <> metavar "SOCKET"
+         <> help "Path to node socket." )
+      <*> strOption
+          ( long "database"
+         <> metavar "DATABASE"
+         <> help "Path to database." )
+
+-- We only care about the mainnet
+slotConfig :: SlotConfig
+slotConfig =
+  SlotConfig
+    { scSlotZeroTime = 1596059091000
+    , scSlotLength   = 1000
+    }
+
+networkId :: NetworkId
+networkId = Mainnet
+
+-- We don't generally need to sync blocks earlier than the Goguen era (other than
+-- testing for memory leaks) so we may want to start synchronising from a slot that
+-- is closer to Goguen era.
```

Comment:
For datums, yes. But some indexers will want to index from start (like the utxosByAddress indexer)

Reply:
I will remove it when I will add the other indexer.